source/cgal/intersections.cc

```diff
@@ -0,0 +1,737 @@
+    {
+#  if defined(CGAL_GEQ_515)
+
+      std::array<CGALPoint3, 4> pts0;
+      std::array<CGALPoint3, 2> pts1;
+      std::transform(
+        first_simplex.begin(),
+        first_simplex.end(),
+        pts0.begin(),
+        [&](const Point<3> &p) {
+          return CGALWrappers::dealii_point_to_cgal_point<CGALPoint3>(p);
+        });
+
+
+      std::transform(
+        second_simplex.begin(),
+        second_simplex.end(),
+        pts1.begin(),
+        [&](const Point<3> &p) {
+          return CGALWrappers::dealii_point_to_cgal_point<CGALPoint3>(p);
+        });
+
+      CGALTetra    tetra{pts0[0], pts0[1], pts0[2], pts0[3]};
+      CGALSegment3 segm{pts1[0], pts1[1]};
+      return CGAL::intersection(segm, tetra);
+
+#  else
+
+      Assert(
+        false,
+        ExcMessage(
+          "This function requires a version of CGAL greater or equal than 5.1.5."));
+      (void)first_simplex;
+      (void)second_simplex;
+      return {};
+#  endif
+    }
+
+
+
+    // tetra, triangle
+    boost::optional<boost::variant<CGALPoint3,
+                                   CGALSegment3,
+                                   CGALTriangle3,
+                                   std::vector<CGALPoint3>>>
+    compute_intersection(const std::array<Point<3>, 3> &first_simplex,
+                         const std::array<Point<3>, 4> &second_simplex)
+    {
+#  if defined(CGAL_GEQ_515)
+      std::array<CGALPoint3, 4> pts0;
+      std::array<CGALPoint3, 3> pts1;
+      std::transform(
+        first_simplex.begin(),
+        first_simplex.end(),
+        pts0.begin(),
+        [&](const Point<3> &p) {
+          return CGALWrappers::dealii_point_to_cgal_point<CGALPoint3>(p);
+        });
+
+      std::transform(
+        second_simplex.begin(),
+        second_simplex.end(),
+        pts1.begin(),
+        [&](const Point<3> &p) {
+          return CGALWrappers::dealii_point_to_cgal_point<CGALPoint3>(p);
+        });
+      CGALTetra     tetra{pts0[0], pts0[1], pts0[2], pts0[3]};
+      CGALTriangle3 triangle{pts1[0], pts1[1], pts1[2]};
+      return CGAL::intersection(triangle, tetra);
+#  else
+
+      Assert(
+        false,
+        ExcMessage(
+          "This function requires a version of CGAL greater or equal than 5.1.5."));
+      (void)first_simplex;
+      (void)second_simplex;
+      return {};
+#  endif
+    }
+  } // namespace internal
+
+
+
+  // Specialization for quads
+  template <>
+  std::vector<std::array<Point<2>, 3>>
+  compute_intersection_of_cells<2, 2, 2>(
+    const typename Triangulation<2, 2>::cell_iterator &cell0,
+    const typename Triangulation<2, 2>::cell_iterator &cell1,
+    const Mapping<2, 2> &                              mapping0,
+    const Mapping<2, 2> &                              mapping1,
+    const double                                       tol)
+  {
+    std::array<Point<2>, 4> vertices0, vertices1;
+    std::copy_n(mapping0.get_vertices(cell0).begin(), 4, vertices0.begin());
```

Comment:
Is it possible to check that mappings are of degree one? I think this is implyed if only vertices of cells are used. Do you know how high order mappings would have to be treated?

Reply:
Yes, this is limited to what we have called "affine cells" in the documentation of this function. I don't know how one should work with high order mappings, as I don't see how it's possible to instruct CGAL to construct and then intersect polygons with curves.

Reply:
Ok, I see. Thanks for clarification :)